File: academic-system-java/src/main/java/org/example/academic/system/model/AcademicClass.java
package org.example.academic.system.model;

import jakarta.validation.constraints.NotBlank;
import org.example.academic.system.exception.AcademicSystemException;
import org.example.academic.system.validation.DomainValidator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class AcademicClass {
    @NotBlank(message = "O codigo da turma nao pode estar vazio.")
    private final String code;

    @NotBlank(message = "O titulo da turma nao pode estar vazio.")
    private final String title;

    private final List<Assessment> assessments;

    // Cria uma turma com código e título, removendo espaços extras e validando os campos
    public AcademicClass(String code, String title) {
        this.code = code == null ? null : code.trim();
        this.title = title == null ? null : title.trim();
        this.assessments = new ArrayList<>();

        DomainValidator.validate(this);
    }

    // Retorna o código da turma
    public String getCode() {
        return code;
    }

    // Retorna o título da turma
    public String getTitle() {
        return title;
    }

    // Adiciona uma avaliação à turma, rejeitando valores nulos
    public void addAssessment(Assessment assessment) {
        if (assessment == null) {
            throw new AcademicSystemException("A avaliacao nao pode ser nula.");
        }

        assessments.add(assessment);
    }

    // Retorna a lista de avaliações como visão imutável
    public List<Assessment> getAssessments() {
        return Collections.unmodifiableList(assessments);
    }

    // Representação textual da turma no formato "código - título"
    @Override
    public String toString() {
        return code + " - " + title;
    }

    // Duas turmas são iguais se tiverem o mesmo código
    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }

        if (!(object instanceof AcademicClass that)) {
            return false;
        }

        return Objects.equals(code, that.code);
    }

    // Hash baseado apenas no código da turma para consistência com equals
    @Override
    public int hashCode() {
        return Objects.hash(code);
    }
}

File: academic-system-python/academic_system/model/academic_class.py
from typing import List

from .assessment import Assessment
from ..exception.academic_system_exception import AcademicSystemException


class AcademicClass:
    def __init__(self, code: str, title: str):
        if not code or not code.strip():
            raise AcademicSystemException("O codigo da turma nao pode estar vazio.")
        if not title or not title.strip():
            raise AcademicSystemException("O titulo da turma nao pode estar vazio.")
        self._code = code.strip()
        self._title = title.strip()
        self._assessments: List[Assessment] = []

    @property
    def code(self) -> str:
        return self._code

    @property
    def title(self) -> str:
        return self._title

    @property
    def assessments(self) -> List[Assessment]:
        return list(self._assessments)

    def add_assessment(self, assessment: Assessment) -> None:
        if assessment is None:
            raise AcademicSystemException("Assessment cannot be None")
        self._assessments.append(assessment)

    def __eq__(self, other) -> bool:
        if not isinstance(other, AcademicClass):
            return False
        return self._code == other._code

    def __hash__(self) -> int:
        return hash(self._code)

    def __str__(self) -> str:
        return f"AcademicClass(code={self._code}, title={self._title})"
